### PR TITLE
[BIP-32] Fixed incorrect HDPath for private key

### PR DESCRIPTION
**Pull request by milczarekIT**
Fixed https://github.com/bitcoinj/bitcoinj/issues/1990

### CODE REVIEW

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java [hunk 1]
```diff
@@ -64,56 +68,80 @@ public void testChildKeyDerivation() throws Exception {
             //////////////////////////////////////////////////////////////////////////
             // Start with an extended PRIVATE key
             DeterministicKey ekprv = HDKeyDerivation.createMasterPrivKeyFromBytes(priv, chain);
+            assertEquals("m", ekprv.getPath().toString());
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:71]: Thanks for adding these asserts!

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java [hunk 2]
```diff
@@ -37,6 +37,10 @@
     private static final int HDW_CHAIN_EXTERNAL = 0;
     private static final int HDW_CHAIN_INTERNAL = 1;
 
+    private static String hexEncodePub(DeterministicKey pubKey) {
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:40]: Can you remove the moving of this method from this PR, to keep it clean?

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java [hunk 3]
```diff
@@ -64,56 +68,80 @@ public void testChildKeyDerivation() throws Exception {
...
             DeterministicKey ekprv_0_IN_1 = HDKeyDerivation.deriveChildKey(ekprv_0_IN, 1);
+            assertEquals("m/0/1/1", ekprv_0_IN_1.getPath().toString());
             DeterministicKey ekprv_0_IN_2 = HDKeyDerivation.deriveChildKey(ekprv_0_IN, 2);
+            assertEquals("m/0/1/2", ekprv_0_IN_2.getPath().toString());
 
             // Now add a few more addresses with very large indices
             DeterministicKey ekprv_1_IN = HDKeyDerivation.deriveChildKey(ekprv_1, HDW_CHAIN_INTERNAL);
+            assertEquals("m/1/1", ekprv_1_IN.getPath().toString());
             DeterministicKey ekprv_1_IN_4095 = HDKeyDerivation.deriveChildKey(ekprv_1_IN, 4095);
-//            ExtendedHierarchicKey ekprv_1_IN_4bil = HDKeyDerivation.deriveChildKey(ekprv_1_IN, 4294967295L);
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:89]: Why did you remove this line…

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java [hunk 4]
```diff
@@ -64,56 +68,80 @@ public void testChildKeyDerivation() throws Exception {
...
             DeterministicKey ekpub_0_IN_1 = HDKeyDerivation.deriveChildKey(ekpub_0_IN, 1);
+            assertEquals("M/0/1/1", ekpub_0_IN_1.getPath().toString());
             DeterministicKey ekpub_0_IN_2 = HDKeyDerivation.deriveChildKey(ekpub_0_IN, 2);
+            assertEquals("M/0/1/2", ekpub_0_IN_2.getPath().toString());
 
             // Now add a few more addresses with very large indices
             DeterministicKey ekpub_1_IN = HDKeyDerivation.deriveChildKey(ekpub_1, HDW_CHAIN_INTERNAL);
+            assertEquals("M/1/1", ekpub_1_IN.getPath().toString());
             DeterministicKey ekpub_1_IN_4095 = HDKeyDerivation.deriveChildKey(ekpub_1_IN, 4095);
-//            ExtendedHierarchicKey ekpub_1_IN_4bil = HDKeyDerivation.deriveChildKey(ekpub_1_IN, 4294967295L);
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:116]: …and this line?

File: core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java [hunk 5]
```diff
@@ -80,21 +80,11 @@ public static DeterministicKey createMasterPrivateKey(byte[] seed) throws HDDeri
...
-    public static DeterministicKey createMasterPrivKeyFromBytes(byte[] privKeyBytes, byte[] chainCode)
-            throws HDDerivationException {
-        // childNumberPath is an empty list because we are creating the root key.
-        return createMasterPrivKeyFromBytes(privKeyBytes, chainCode, HDPath.M());
-    }
-
-    /**
-     * @throws HDDerivationException if privKeyBytes is invalid (not between 0 and n inclusive).
-     */
-    public static DeterministicKey createMasterPrivKeyFromBytes(byte[] privKeyBytes, byte[] chainCode,
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java:92]: Why not keep a variant of this constructor with the path?

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java (hunk 3 above)
Comment (milczarekIT) [reply core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:89]: It was commented an old code. Git repository is not a place for keeping old code in comments.

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java (hunk 4 above)
Comment (milczarekIT) [reply core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:116]: It was commented an old code. Git repository is not a place for keeping old code in comments.

File: core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java (hunk 5 above)
Comment (milczarekIT) [reply core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java:92]: Master private key always has path: 'm'

See: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

File: core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java (hunk 4 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/crypto/ChildKeyDerivationTest.java:116]: That's right, I overlooked this.

File: core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java (hunk 5 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/crypto/HDKeyDerivation.java:92]: Indeed, that makes sense.

**Comment by schildbach** [comment]
I decided to remove the commented tests in a separate commit, because it actually consisted of 3 commented out lines instead of the two you removed.

Then I rebased your PR and merged. Thanks again!
